Renderer/RenderProcess.php

```diff
@@ -244,28 +254,50 @@ private function wrapTextWithMarker (array $openMarkupIndexes, int $numberOfClos
     private function renderOpeningMarkup (array $markup) : string
     {
         $tagName = $markup[0];
-        $attributes = $markup[1] ?? null;
+        $flatAttributes = $markup[1] ?? null;
 
-        // if only an opening tag
-        if (null === $attributes)
+        // if only an opening tag or invalid attributes
+        if (!\is_array($flatAttributes))
         {
             return "<{$tagName}>";
         }
 
-        $numberOfAttributes = count($attributes) / 2;
-        $renderedAttributes = [];
-        for ($i = 0; $i < $numberOfAttributes; $i++)
+        // parse attribute list to a structured array
+        $attributes = $this->parseFlatAttributes($flatAttributes);
+
+        foreach ($this->markupAttributesVisitors as $visitor)
         {
-            $name = $attributes[$i * 2];
-            $value = $attributes[($i * 2) + 1];
+            $transformed = $visitor->transform($tagName, $attributes);
+
+            if (null !== $transformed)
+            {
+                $attributes = $transformed;
+                break;
+            }
+        }
+
+        foreach ($attributes as $key => $value)
+        {
+            // skip `false` and `null` attributes
+            if ($value === false || $value === null)
+            {
+                continue;
+            }
+
+            // for `true` attributes, just use the key
+            if ($value === true)
```

Comment:
What about empty attributes such as `href=""`?

Do we need some special treatment there as well?

Reply:
You can just do `["href", ""]` and it will be rendered as `href=""`.